File: sdk/go/dispatch.go
// Package dispatch is the Go client for wave Dispatch — route each request to the cheapest capable
// model (local-first, escalate to your frontier model only when needed). Your keys + infra stay yours;
// the API returns a routing decision (and, if your plan enables it, the edge-local answer).
package dispatch

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// DefaultEndpoint is the hosted wave Dispatch edge.
const DefaultEndpoint = "https://dispatch.wave.online"

// Client calls the wave Dispatch edge API.
type Client struct {
	Endpoint string       // defaults to DefaultEndpoint
	License  string       // Bearer license key (wv_...); empty = x402 pay-per-use
	HTTP     *http.Client // defaults to a 30s client
}

// New returns a Client for the hosted edge with the given license key.
func New(license string) *Client {
	return &Client{Endpoint: DefaultEndpoint, License: license, HTTP: &http.Client{Timeout: 30 * time.Second}}
}

// Decision is the routing result.
type Decision struct {
	Route       string  `json:"route"`
	Probability float64 `json:"probability"`
	Margin      float64 `json:"margin"`
	DecidedAt   string  `json:"decided_at"`
	Forward     bool    `json:"forward"`              // true => escalate to your frontier model
	Answer      string  `json:"answer,omitempty"`     // present when Execute is requested + enabled
	ExecutedBy  string  `json:"executed_by,omitempty"`
	Tier        string  `json:"tier,omitempty"`
}

// Route classifies a prompt (no execution).
func (c *Client) Route(ctx context.Context, prompt string) (*Decision, error) {
	return c.post(ctx, map[string]any{"prompt": prompt})
}

// Execute classifies and runs the request on the edge if your plan allows it.
func (c *Client) Execute(ctx context.Context, prompt string) (*Decision, error) {
	return c.post(ctx, map[string]any{"prompt": prompt, "execute": true})
}

// RouteVector classifies a pre-computed 768-d embedding (matmul-only — cheapest + fastest path).
func (c *Client) RouteVector(ctx context.Context, vec []float64) (*Decision, error) {
	return c.post(ctx, map[string]any{"vector": vec})
}

func (c *Client) post(ctx context.Context, body map[string]any) (*Decision, error) {
	ep := c.Endpoint
	if ep == "" {
		ep = DefaultEndpoint
	}
	b, _ := json.Marshal(body)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, ep+"/", bytes.NewReader(b))
	if err != nil {
		return nil, err
	}
	req.Header.Set("content-type", "application/json")
	if c.License != "" {
		req.Header.Set("authorization", "Bearer "+c.License)
	}
	hc := c.HTTP
	if hc == nil {
		hc = http.DefaultClient
	}
	resp, err := hc.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode == http.StatusPaymentRequired {
		return nil, fmt.Errorf("dispatch: 402 payment required (x402) — pay and retry, or set a license")
	}
	if resp.StatusCode != http.StatusOK {
		var e struct {
			Error string `json:"error"`
		}
		_ = json.NewDecoder(resp.Body).Decode(&e)
		return nil, fmt.Errorf("dispatch: status %d: %s", resp.StatusCode, e.Error)
	}
	var d Decision
	if err := json.NewDecoder(resp.Body).Decode(&d); err != nil {
		return nil, err
	}
	return &d, nil
}

File: sdk/ruby/dispatch.rb
# wave Dispatch — Ruby client. Route each request to the cheapest capable model (local-first;
# escalate to your frontier model only when needed). Your keys + infra stay yours.
require "net/http"
require "json"
require "uri"

module Dispatch
  DEFAULT_ENDPOINT = "https://dispatch.wave.online".freeze

  class Client
    def initialize(license = nil, endpoint: DEFAULT_ENDPOINT)
      @license = license
      @endpoint = endpoint
    end

    # Classify a prompt (no execution). Returns a Hash: route, probability, margin, forward, ...
    def route(prompt)
      post("prompt" => prompt)
    end

    # Classify and run on the edge (if your plan enables it).
    def execute(prompt)
      post("prompt" => prompt, "execute" => true)
    end

    # Classify a pre-computed 768-d embedding (matmul-only: cheapest + fastest).
    def route_vector(vec)
      post("vector" => vec)
    end

    private

    def post(body)
      uri = URI(@endpoint.to_s.chomp("/") + "/")
      req = Net::HTTP::Post.new(uri)
      req["content-type"] = "application/json"
      req["authorization"] = "Bearer #{@license}" if @license && !@license.empty?
      req.body = JSON.generate(body)
      res = Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https") { |h| h.request(req) }
      code = res.code.to_i
      return JSON.parse(res.body) if code == 200
      raise "dispatch: 402 payment required (x402) — pay and retry, or set a license" if code == 402

      err = (JSON.parse(res.body)["error"] rescue res.body)
      raise "dispatch: status #{code}: #{err}"
    end
  end
end

if __FILE__ == $PROGRAM_NAME
  c = Dispatch::Client.new(ENV["WAVE_LICENSE"])
  d = c.route("find the auth handler")
  puts "route=#{d['route']} prob=#{d['probability']} forward=#{d['forward']}"
end
